platform/pc/boot/stage1.s

```diff
@@ -75,11 +75,42 @@ dap:
         .lba          dq 1
 
 
+%ifdef DEBUG
+bits 16
+%include "debug.inc"
+bits 32
+%endif
+
 readsectors:
+        mov eax, [dap.sector_count]
+        cmp eax, 0x80
+        jle loop
+        mov ecx, 0x80
+        mov [dap.sector_count], ecx
+loop:
         mov si, dap
         mov ah, 0x42
         mov dl, 0x80
         int 0x13
+%ifdef DEBUG
+        jc sector_read_error
```

Comment:
If I try this after removing the ifdef DEBUG, it reports an error. Indeed, the int 13h is returning with CF set and error code (AH) set to 0x09, which is "Attempt to DMA over 64kb boundary."

It looks like the jle above is not doing what is expected and is jumping to loop even though the sector_count is 0xff (in my test case). Apparently, this is because the 'cmp' above is doing only a signed byte comparison, and both 0x80 and 0xff are being compared as negative values. Changing the immediate from 0x80 to 0x0080 seems to do the trick (emits a cmpw). Since these are unsigned values, too, it should technically be jbe instead of jle, though that shouldn't matter in this case.